libp2p/Host.cpp

```diff
@@ -698,56 +698,61 @@ void Host::run(boost::system::error_code const&)
         return;
     }
 
-    if (auto nodeTable = this->nodeTable()) // This again requires x_nodeTable, which is why an additional variable nodeTable is used.
-        nodeTable->processEvents();
+    // Don't need to process node table events (which results in peers being added/dropped) or
+    // perform any p2p functionality if we aren't running p2p
+    if (m_capabilities.size())
```

Comment:
Hmm if everything in this function is not needed without capabilities, probably we should just not schedule it for run, that is skip this https://github.com/ethereum/aleth/blob/b92056768376d42b809e2a80afd3178b2dead29f/libp2p/Host.cpp#L802

Reply:
Well the function deletes the node table, stops the io service, and stops the timer (which is used to schedule future calls to run()): https://github.com/ethereum/aleth/blob/b92056768376d42b809e2a80afd3178b2dead29f/libp2p/Host.cpp#L683-L699

I don't know enough about how boost's io service works and the interaction between the host and the node table to know if it's safe to skip this code or if it can be moved into another function. I'll do some more research.

Reply:
Some more thoughts about this

- I would try to move all this under `if (!m_run)` out of this function, perhaps to `Host::stop()` (the only place that sets `m_run` to `false`)
- `m_ioService.stop()` may be safely called from another thread, I found discussion about this here https://stackoverflow.com/a/4810965/1636308
- `m_nodeTable` should be deleted after `m_ioService.stop()` call anyway to fix https://github.com/ethereum/aleth/issues/5388. Or perhaps it should be not deleted at all when network is stopped (after `io_service` is stopped nothing will be called in `NodeTable` anyway) and we should just leave its destruction to `Host` destructor
- I don't see any good reason to destroy `m_timer` when network is stopped. Maybe these were some hacks to handle the edge case when `Host::stop()` is being called while its starting is still in progress: this comment points to it https://github.com/ethereum/aleth/blob/60a1075455feaf497cc8313b829f851fd3692574/libp2p/Host.cpp#L753-L755
but it seems weird to use `m_timer` for this